crypto/context.c

```diff
@@ -145,6 +146,16 @@ OPENSSL_CTX *OPENSSL_CTX_new(void)
     return ctx;
 }
 
+#ifndef FIPS_MODE
+int OPENSSL_CTX_load_config(OPENSSL_CTX *ctx, const char *config_file)
+{
+    if (!OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL))
+        return 0;
```

Comment:
Why is this needed? It seems unnecessary?

Reply:
Because there are dependancies which we want to ensure are called once before loading..
i.e: openssl_config_int() calls OPENSSL_load_builtin_modules()

Take out the line and run evp_fetch_prov_test to see this in action.

Reply:
Ah - right. Thanks, I understand now. However, I think this is misplaced. We should be making these init calls "just in time" IMO. So in this case the problem is we are trying to use built-in modules before they have been loaded. Therefore the function where we try to get the reference to the module should be where we do this just-in-time initialisation, i.e.

```` diff
diff --git a/crypto/conf/conf_mod.c b/crypto/conf/conf_mod.c
index 21d3b009bf..5c781ed33b 100644
--- a/crypto/conf/conf_mod.c
+++ b/crypto/conf/conf_mod.c
@@ -180,6 +180,12 @@ static int module_run(const CONF *cnf, const char *name, const char *value,
     CONF_MODULE *md;
     int ret;
 
+    /* Make sure any built-in modules are loaded before we try to find them */
+    if (!OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL)) {
+        CONFerr(CONF_F_MODULE_RUN, CONF_R_MODULE_INITIALIZATION_ERROR);
+        return -1;
+    }
+
     md = module_find(name);
 
     /* Module not found: try to load DSO */
````

Reply:
That is not going to work.
The OPENSSL_init_crypto() calls 
CONF_modules_load_file() which will call OPENSSL_init_crypto() - which then causes a deadlock.

Reply:
Good point. This actually implies to me that openssl_config_int() should not call OPENSSL_load_builtin_modules() at all. We could then just call OPENSSL_load_builtin_modules() from a RUNONCE in module_run.

Reply:
Ahh ok that seems to work - glad you know what you are doing :)
That seems to clear up the dodgy bits of this PR that made me feel a bit dirty when I added them :)